src/Microsoft.Extensions.Configuration.AzureAppConfiguration/ConfigurationClientProvider.cs

```diff
@@ -0,0 +1,123 @@
+    /// </summary>
+    /// <remarks>
+    /// This class is not thread-safe. Since config provider does not allow multiple network requests at the same time,
+    /// there won't be multiple threads calling this client at the same time.
+    /// </remarks>
+    internal class ConfigurationClientProvider : IConfigurationClientProvider
+    {
+        private readonly IList<ConfigurationClientStatus> _clients;
+
+        public ConfigurationClientProvider(string connectionString, ConfigurationClientOptions clientOptions)
+        {
+            if (string.IsNullOrEmpty(connectionString))
+            {
+                throw new ArgumentNullException(nameof(connectionString));
+            }
+
+            var endpoint = new Uri(ConnectionStringParser.Parse(connectionString, ConnectionStringParser.EndpointSection));
+            var configurationClientStatus = new ConfigurationClientStatus(endpoint, new ConfigurationClient(connectionString, clientOptions));
+            _clients = new List<ConfigurationClientStatus> { configurationClientStatus };
+        }
+
+        public ConfigurationClientProvider(IEnumerable<Uri> endpoints, TokenCredential credential, ConfigurationClientOptions clientOptions)
+        {
+            if (endpoints == null || !endpoints.Any())
+            {
+                throw new ArgumentNullException(nameof(endpoints));
+            }
+
+            _clients = endpoints.Select(endpoint => new ConfigurationClientStatus(endpoint, new ConfigurationClient(endpoint, credential, clientOptions))).ToList();
+        }
+
+        /// <summary>
+        /// Internal constructor; Only used for unit testing.
+        /// </summary>
+        /// <param name="clients"></param>
+        internal ConfigurationClientProvider(IList<ConfigurationClientStatus> clients)
+        {
+            _clients = clients;
+        }
+
+        public IEnumerable<ConfigurationClient> GetClients()
+        {
+            List<ConfigurationClient> clients = new List<ConfigurationClient>();
+
+            foreach(ConfigurationClientStatus configurationClient in _clients)
+            {
+                if (configurationClient.BackoffEndTime <= DateTimeOffset.UtcNow)
+                {
+                    clients.Add(configurationClient.Client);
+                }
+            }
+
+            // If all clients are in the back-off state, try all clients anyways.
+            if (!clients.Any())
+            {
+                clients.AddRange(_clients.Select(c => c.Client));
+            }
+
+            return clients;
+        }
+
+        public void UpdateClientStatus(ConfigurationClient client, bool successful)
+        {
+            if (client == null)
+            {
+                throw new ArgumentNullException(nameof(client));
+            }
+
+            ConfigurationClientStatus clientWrapper = _clients.First(c => c.Client.Equals(client));
```

Comment:
Do you like to change `ConfigurationClientStatus` to `ConfigurationClientWrapper`?

Reply:
I have a double mind here. The class isn't just wrapping the `ConfigurationClient`, but it is also maintaining the status of it.
I think `ConfigurationClientStatus` is more indicative of what it is doing, unless there is another name that suggested it is a wrapper as well as maintains status.

Reply:
Exactly like you said, this class wraps the _client itself_ AND its _status_ (well, and some other properties). Calling it just status does not accurately indicate what it does. Actually, you use it just like a client in many places. The status and other properties (like endpoint) are just attributes of the client. Imagining if the `ConfigurationClient` exposed the endpoint property and had data bag for extra data storage, we would not need this class. "Wrapper" may not be the best name, but it accurately describes what we have here.

Reply:
How do you feel about `ConfigurationClientStatusWrapper`?

Reply:
It's not a wrapper of status.